crates/parser/src/format/json.rs

```diff
@@ -21,7 +22,120 @@ impl Parser for JsonParser {
             .transcode_non_utf8(None, 32)?
             .into_buffered_stream(8192);
         let deser = serde_json::de::Deserializer::from_reader(input).into_iter();
-        let wrapped = deser.map(|r| r.map_err(Into::into));
+        let wrapped = JsonIter {
+            inner: Box::new(deser),
+            current_array: None,
+        };
         Ok(Box::new(wrapped))
     }
 }
+
+struct JsonIter {
+    inner: Box<dyn Iterator<Item = Result<Value, serde_json::Error>>>,
+    current_array: Option<std::vec::IntoIter<Value>>,
+}
+
+impl Iterator for JsonIter {
+    type Item = Result<Value, ParseError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(array) = self.current_array.as_mut() {
+            if let Some(elem) = array.next() {
+                if elem.is_object() {
+                    return Some(Ok(elem));
+                } else {
+                    return Some(Err(ParseError::Parse(InvalidJsonType(elem).into())));
+                }
+            }
+        }
+        let next_elem = self.inner.next()?;
+        match next_elem {
+            Ok(Value::Array(array)) => {
+                self.current_array = Some(array.into_iter());
+                self.next()
+            }
+            Ok(other) if other.is_object() => Some(Ok(other)),
+            Ok(other) => Some(Err(ParseError::Parse(InvalidJsonType(other).into()))),
+            Err(e) => Some(Err(e.into())),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("The JSON parser only supports objects or arrays of objects, found scalar value: '{0}'")]
```

Comment:
s/scalar value/value/? Just because it sounds silly if given an input like `[[1,2],[3,4]` to say "scalar value: `[1,2]`".